File: go/breadth-first-search/intro.go
package main

type TreeNode struct {
	Left  *TreeNode
	Right *TreeNode
	Val   int
}

/*
	BFS:
	Uses queue to keep track of nodes it needs to visit
*/

func bfs(root *TreeNode) []int {
	if root == nil {
		return []int{}
	}

	result := []int{}

	// Start by putting the first root in the queue
	queue := []*TreeNode{root}

	for len(queue) > 0 {
		currNode := queue[0]
		queue = queue[1:] // remove item at the beginning of the queue

		result = append(result, currNode.Val)

		if currNode.Left != nil {
			queue = append(queue, currNode.Left)
		}

		if currNode.Right != nil {
			queue = append(queue, currNode.Right)
		}
	}

	return result
}

File: breadth-first-search/intro.py
from collections import deque


def dfs(root):
    if not root:
        return []

    result = []

    queue = deque([root])

    while queue:
        curr_node = queue.popleft()

        result.append(curr_node.val)

        if curr_node.left:
            queue.append(curr_node.left)

        if curr_node.right:
            queue.append(curr_node.right)

    return result
